core/client/fs/src/main/java/alluxio/client/block/policy/options/CreateOptions.java

```diff
@@ -21,21 +25,58 @@
  */
 @PublicApi
 public final class CreateOptions {
+  private AlluxioConfiguration mAlluxioConf;
   private String mLocationPolicyClassName;
   private int mDeterministicHashPolicyNumShards;
+  private TieredIdentity mTieredIdentity;
+  private long mBlockCapacityReserved;
+  private String mSpecificWorker;
 
   /**
+   * Creates a new instance of options for {@link alluxio.client.block.policy.BlockLocationPolicy}.
+   *
+   * Note that the setter for configuration is explicity excluded from this class. To override
+   * values in the options after calling {@code defaults()} programmers should use the respective
+   * setter.
+   *
+   * @param alluxioConf Alluxio configuration
    * @return the default {@link CreateOptions}
    */
-  public static CreateOptions defaults() {
-    return new CreateOptions();
+  public static CreateOptions defaults(AlluxioConfiguration alluxioConf) {
+    return new CreateOptions(alluxioConf);
   }
 
   /**
    * Creates a new instance with defaults.
+   *
+   * @param alluxioConf Alluxio configuration
    */
-  private CreateOptions() {
-    mDeterministicHashPolicyNumShards = 1;
+  private CreateOptions(AlluxioConfiguration alluxioConf) {
```

Comment:
Do we need this class or could we just use the configuration instead?

Reply:
Right now it's used because client RPCs can override the configuration value for numShards in DeterministicHashPolicy. If we wanted to add the ability to modify other policy parameters in the future via RPCs, this class allows that.

Otherwise, I'm not sure how I would want implement overriding the hash shards if we didn't have this class.

Reply:
Could you elaborate on what you mean by `client RPCs can override the configuration value for numShards`?

Reply:
https://github.com/Alluxio/alluxio/blob/master/core/transport/src/grpc/file_system_master.proto#L66 - a client can override whatever configuration value is set in the master client based on an RPC's `OpenFilePOptions`

Reply:
Given the assumption that we remove CreateOptions and replace the constructor with a combination of configuration/OpenFilePOptions - Currently, the deterministic hashing policy is the only policy which allows values to be set in the RPC options. If other policies were to be added in the future, would they also be put into OpenFilePOptions? Right now none of the other policies rely on RPC options, rather configuration values from the master.

Keeping `CreateOptions` makes instantiating simpler because it provides a uniform interface for us to create instances of the BlockPolicies. Otherwise, some instances rely on the RPC options, others for configuration. I can imagine a case where a policy may rely on one or the other, or both, for a policy to be created.